src/main/java/Parser.java

```diff
@@ -0,0 +1,324 @@
+import java.time.LocalDate;
+import java.util.ArrayList;
+
+/**
+ * Encapsulates a Parser
+ * Deals with making sense of the user command
+ */
+
+public class Parser {
+
+    protected Storage storage;
+    protected Ui ui;
+    protected TaskList tasks;
+
+    /**
+     * Constructor
+     *
+     * @param storage is the storage that stores the data file
+     * @param ui is the user interface used with duke
+     * @param list is the TaskList that will be updated
+     */
+    public Parser(Storage storage, Ui ui, TaskList list) {
+        this.storage = storage;
+        this.ui = ui;
+        this.tasks = list;
+    }
+
+    /**
+     * Processes the user's inputs
+     *
+     * @param input is the user's input in String form
+     * @throws DukeException
+     */
+    protected String processInput(String input) throws DukeException {
+        int stringLength = input.length();
+        assert stringLength >= 1 : "Input cannot be empty";
+        if (input.equals("list")) {
+            return this.provideList();
+        } else if (input.startsWith("done")) {
+            return this.markAsDone(input);
+        } else if (input.startsWith("todo") || input.startsWith("deadline") || input.startsWith("event")) {
+            return this.newTaskEntry(input);
+        } else if (input.equals("bye")) {
+            return this.end();
+        } else if (input.startsWith("delete")) {
+            return this.delete(input);
+        } else if (input.startsWith("find")) {
+            return this.find(input);
+        } else if (input.equals("remove duplicates")) {
+            return this.removeDupes();
+        } else {
+            return new InputNotRecognisedException().toString();
+        }
+    }
+
+    /**
+     * Compares two tasks
+     *
+     * @return a boolean denoting if they are equal
+     */
+    public boolean areEqualTasks(Task t1, Task t2) {
+        if (t1 instanceof Todo && t2 instanceof Todo) {
+            Todo thisTask = (Todo) t1;
+            Todo otherTask = (Todo) t2;
+            return t1.equals(t2);
+        } else if (t1 instanceof Deadline && t2 instanceof Deadline) {
+            Deadline thisTask = (Deadline) t1;
+            Deadline otherTask = (Deadline) t2;
+            return t1.equals(t2);
+        } else if (t1 instanceof Event && t2 instanceof Event) {
+            Event thisTask = (Event) t1;
+            Event otherTask = (Event) t2;
+            return t1.equals(t2);
+        } else {
+            return false;
+        }
```

Comment:
Actually, since you have already overwritten the `equals` method in your Todo/Deadline/Event, wouldn't the checks for them being of the same instance be handled in the overwritten `equals` method? I'm not 100% sure and may be wrong here.